### PR TITLE
My own summer neural network post for review

### PR DESCRIPTION
**Pull request by juhofriman**
Tälläinen hauska lomaprojekti tuli tunkattua ja iski sitten kirjoitusripulikin asian tiimoilta.

### CODE REVIEW

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+You might ask why I didn't just pick a simple game from the internet? It's because I wanted to keep the machine learning part as simple as possible, and thought that the easiest way to achieve this is to have full control of the game mechanics.
+
+## The Game - Endless Flighter
+
+I poured another cup of joe, and rolled up some [Parcel](https://parceljs.org/) and [Three.js](https://threejs.org/). After couple of hours I had [Endless Flighter](https://github.com/juhofriman/endless-flighter) all set up and running on my browser.
+
+![Endless Flighter Human](/img/endless-flighter-human.gif)
+
+Almost as simple as possible game in which you can steer left or right and you should avoid obstacles. I did not even implement collision detection to my game, because I realised that it is not actually needed and it only makes things more complex. This way my learning data collection and testing the model is much more straight forward.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Almost as simple as possible game in which you can steer left or right and you should avoid obstacles. I did not even implement collision detection, because I realized that it is not actually needed and it only makes things more complex. This way my learning data collection and testing the model is much more straightforward.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+## The AI Infrastructure
+
+The rough idea I learned from those Python Plays GTA videos is this:
+
+1. Implement capturing mechanism that takes screenshots, preprocesses them and registers pressed keys at the time of screenshot
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
1. Implement capturing mechanism that takes screenshots, preprocesses them and registers pressed keys at the time of the screenshot
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+
+![Rough Architecture](/img/endless-flighter-architecture.png)
+
+All the code used in my experiment is available [here](https://github.com/juhofriman/endless-flighter-ai).
+
+### 1. Capturing the data
+
+So the first thing to do is to implement a program, which captures whole lotta screenshots of manual flying and tracks keypresses.
+
+This is actually pretty easy to do. Just place your game to a certain position in your screen and use `ImageGrab` or similar to grab that image.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
This is actually pretty easy to do. Just place your game to a certain position in your screen and use `ImageGrab` or a similar program to grab that image.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+Then you probably need to grayscale the image because colour images are huge compared to grayscale images. If you have more complex game, you probably should also apply edge detection and other image processing techniques to images in order to get them more simple for the neural network. Grayscaling seemed to be enough for my simple game. It's also important to make learning data as small as possible, so don't use 600x600 images but instead scale images down to 60x60 or so - it should be enough for the neural network to figure things out.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Then you probably need to grayscale the image because colour images are huge compared to grayscale images. If you have more complex game, you probably should also apply edge detection and other image processing techniques to the images in order to render them more simple for the neural network. Grayscaling seemed to be enough for my simple game. It's also important to make the learning data as small as possible, so don't use 600x600 images but instead scale images down to 60x60 or so - it should be enough for the neural network to figure things out.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  ...
+  [image_matrix, keyboard]
+]
+```
+
+![Training data replayed](/img/endless-flighter-training-data.gif)
+
+Above is a visualisation of the learning data replayed. On the terminal, you can see which buttons were pressed, and on the window is the corresponding screen.
+
+I got pretty low frame rate of couple of images per second, but it turned out that this frame rate is actually enough for my simple Endless Flighter game. Note that images in learning data are cropped so that the bars behind player are not visible. This originates from my assumption, that the bars already passed do not matter anymore. When you play the game, you glaze your vision to the upper part of the screen.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
I got pretty low frame rate of couple of images per second, but it turned out that this frame rate is actually enough for my simple Endless Flighter game. Note that images in the learning data are cropped so that the bars behind the player are not visible. This originates from my assumption, that the bars already passed do not matter anymore. When you play the game, you glaze your vision to the upper part of the screen.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+### 2. Capture Lots of Data by Flying Manually
+
+I wanted to make this work end-to-end with minimal training data set, because I did not want to spent hours flying the most dull game in the history for hours and then see that it didn't work and the data is corrupt somehow. So I decided just to fly for a couple of minutes and try to achieve end-to-end solution which will probably be really bad because of the small amount of data.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
I wanted to make this work end-to-end with a minimal training data set, because I did not want to spent hours flying the most dull game in the and then see that it didn't work and the data is corrupt somehow. So I decided just to fly for a couple of minutes and try to achieve end-to-end solution which will probably be really bad because of the small amount of data.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+### 3. Balance the Data
+
+I learned, that overfitting is a common problem with neural networks. Overfitting means that you have too much of certain values in the learning data and then the network would pretty much always do that. In this example flying straight is way more common than steering and thus you need to drop "flying straight" examples to mach the steering examples. The goal is to get roughly the same amount of targets (keyboard states) to learning data.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]: > and then the network would pretty much always do that.
Do what?

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+
+One of the most interesting things is that the ordering of the examples do not matter. Most of the people would intuitively think that playing this sort of game is contextual, meaning that previous examples affect the current status. But this is not how neural networks work, as neural network do not "see" the previous images. Learning data is shuffled, and replaying balanced learning data visually does not make any sense for human anymore - it's just a collection of images and corresponding keyboard states in random order.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
One of the most interesting things is that the ordering of the examples do not matter. Most of the people would intuitively think that playing this sort of game is contextual, meaning that previous examples affect the current status. But this is not how neural networks work, as neural network do not "see" the previous images. Learning data is shuffled, and replaying the balanced learning data visually does not make any sense for a human anymore - it's just a collection of images and corresponding keyboard states in random order.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+I also made a small mistake in the learning data capturing stage, because I captured keyboard presses in a form of an array of `[a_pressed, s_pressed]`:
+
+```
+[0, 0] # nothing pressed
+[1, 0] # a pressed
+[0, 1] # s pressed
+```
+
+[TensorFlow](https://www.tensorflow.org/) produces *predictions* for the probability of the output for input, so you would get back an array of two elements. So for an screenshot, you would get back something like `[0.493, 0.984]` which would suggest in out example to press the `s` button. Usually you select the action with highest probability and this would mean that our self playing agent would *always* turn to some direction. To overcome this, I just simply mapped the data as following:
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]: > in out example

en ummarra

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+        keyboard.release('a')
+    elif moves == [0, 0, 1.0]:
+        keyboard.release('a')
+        keyboard.release('s')
+```
+
+### 6. Let the Agent Play
+
+
+Then I set things up, fired up the processes and to my sheer amazement, the player moved to some directions on the first try! It made mistakes and wasn't particularly intelligent but at least it did something. I guess this was the first (and probably the last) time in my life when something like this seemed to work at least to some extent on first try.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Then I set things up, fired up the processes and to my sheer amazement, the player moved to some directions on the first try! It made mistakes and wasn't particularly intelligent but at least it did something. I guess this was the first (and probably the last) time in my life when something like this seemed to work at least to some extent on the first try.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+
+It still makes mistakes, but at least it's clearly trying to avoid the obstacles.
+
+Then I spent the most dull 40 minutes, and the model got better. It still gets distracted once in a while, but is clearly better with larger data.
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/ZmYCy4eH_Mg" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+I wouldn't jump into self driving car which is using my neural network, but I think I got pretty darn good results for the first try. I guess the main factors to my successful experiment were:
+
+- Game is really simple
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
- The Game is really simple
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+- Visual outlook of the game is really "predictive"
+- It's easy to collect steady and unskewed learning data
+
+## Onward to the World of Machine Learning
+
+Isn't all machine learning done with huge data using massive GPU arrays from the cloud? Probably true, but you can easily start learning this stuff with just your laptop and CPU and you don't need GPU to get started. My largest 40 minutes learning data was just about 40MB in size and I got pretty decent results. The training with TensorFlow took about 5 minutes or so. So go on and try something out! For me, this was probably the most fun project I have ever done.
```
Comment (msantala) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Isn't all machine learning done with huge data using massive GPU arrays from the cloud? Probably true, but you can easily start learning this stuff with just your laptop and CPU and you don't need a GPU to get started. My largest 40 minutes learning data was just about 40MB in size and I got pretty decent results. The training with TensorFlow took about 5 minutes or so. So go on and try something out! For me, this was probably the most fun project I have ever done.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+
+I also made a small mistake in the learning data capturing stage, because I captured keyboard presses in a form of an array of `[a_pressed, s_pressed]`:
+
+```
+[0, 0] # nothing pressed
+[1, 0] # a pressed
+[0, 1] # s pressed
+```
+
+[TensorFlow](https://www.tensorflow.org/) produces *predictions* for the probability of the output for input, so you would get back an array of two elements. So for an screenshot, you would get back something like `[0.493, 0.984]` which would suggest to press the `s` button, because second value in the prediction array - which denominates the probability for `s` key pressed - is larger. Usually you select the action with highest probability and this would mean that our self playing agent would *always* turn to some direction, because there is no probability for going straight. To overcome this, I just simply mapped the data as following:
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
[TensorFlow](https://www.tensorflow.org/) produces *predictions* for the probability of the output for input, so you would get back an array of two elements. So for a screenshot, you would get back something like `[0.493, 0.984]` which would suggest to press the `s` button, because the second value in the prediction array - which denominates the probability for `s` key pressed - is larger. Usually you would select the action with the highest probability and this would mean that our self playing agent would *always* turn to some direction, because there is no probability for going straight. To overcome this, I just simply mapped the data as following:
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+[0, 0] -> [0, 0, 1] # Go straight
+[1, 0] -> [1, 0, 0] # Steer left
+[0, 1] -> [0, 1, 0] # Steer right
+```
+
+From this data, I get predictions in a form of `[go_left_prob, go_right_prob, go_forward_prob]`.
+
+### 4. Enter the TensorFlow
+
+This is the part of my experiment that I have absolutely no knowledge of. I just simply copied the neural network used in GTA examples and I learned only one thing:
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
This is the part of my experiment that I had absolutely no knowledge of. I just simply copied the neural network used in GTA examples and I learned only one thing:
```

:smile:

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+the input and output have shapes which must be defined in the network.
+
+This is the neural network configuration used, but I have absolutely no idea, what it actually does :D
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
This is the neural network configuration used, but I have absolutely no idea what it actually does :D
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+                        max_checkpoints=1, tensorboard_verbose=2, tensorboard_dir='log')
+
+    return model
+```
+
+Then I just ran some code that fits the data to the neural network and to my surprise, it worked and TensorFlow produced the model files!
+
+### 5. Implement the Agent
+
+Implementing the agent is pretty similar to implementing the capturing mechanism, but instead of capturing the current keyboard state, you ask prediction for the image from neural network and use the given result to send a keyboard event.
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Implementing the agent is pretty similar to implementing the capturing mechanism, but instead of capturing the current keyboard state, you ask for a prediction for the image from the neural network and use the given result to send a keyboard event.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+
+And I was like:
+
+![Giphy](https://media.giphy.com/media/Ki9ZNTNS7aC9q/giphy.gif)
+
+**Mission accomplished!**
+
+It still makes mistakes, but at least it's clearly trying to avoid the obstacles.
+
+Then I spent the most dull 40 minutes, and the model got better. It still gets distracted once in a while, but is clearly better with larger data.
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Then I spent the most dull 40 minutes, and the model got better. It still gets distracted once in a while, but is clearly better with more data.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/ZmYCy4eH_Mg" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+I wouldn't jump into self driving car which is using my neural network, but I think I got pretty darn good results for the first try. I guess the main factors to my successful experiment were:
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
I wouldn't jump into a self driving car which is using my neural network, but I think I got pretty darn good results for the first try. I guess the main factors to my successful experiment were:
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+
+- The game is really simple
+- Visual outlook of the game is really "predictive"
+- It's easy to collect steady and unskewed learning data
+
+## Onward to the World of Machine Learning
+
+Isn't all machine learning done with huge data using massive GPU arrays from the cloud? Probably true, but you can easily start learning this stuff with just your laptop and CPU and you don't need a GPU to get started. My largest 40 minutes learning data was just about 40MB in size and I got pretty decent results. The training with TensorFlow took about 5 minutes or so. So go on and try something out! For me, this was probably the most fun project I have ever done.
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
Isn't all machine learning done with huge data sets, using massive GPU arrays from the cloud? Probably true, but you can easily start learning this stuff with just your laptop and CPU and you don't need a GPU to get started. My largest 40 minutes learning data was just about 40MB in size and I got pretty decent results. The training with TensorFlow took about 5 minutes or so. So go on and try something out! For me, this was probably the most fun project I have ever done.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+
+It's almost magical experience when you see your agent playing the game for the first time when you can see something clearly "intelligent" and you haven't programmed a single line of logic yourself, but instead passed loads of data somewhere to make predictions from.
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
It's almost a magical experience when you see your agent playing the game for the first time when you can see something clearly "intelligent" and you haven't programmed a single line of logic yourself, but instead passed loads of data somewhere to make predictions from.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+
+I'm pretty sure, that I am now able to create a program that tries to detect, if an image contains a hotdog.
```
Comment (mikaelpaj) [inline _posts/2019-08-01-my-own-summer-neural-network.md]:
```suggestion
I'm pretty sure, that I am now able to create a program that tries to detect, if an image contains a hot dog.
```

File: _posts/2019-08-01-my-own-summer-neural-network.md (hunk 14 above)
Comment (juhofriman) [reply _posts/2019-08-01-my-own-summer-neural-network.md]: Geniusta
